stubs/WebOb/webob/multidict.pyi

```diff
@@ -0,0 +1,113 @@
+    @overload
+    def pop(self, key: _KT) -> _VT: ...
+    @overload
+    def pop(self, key: _KT, default: _T) -> _VT | _T: ...
+    def popitem(self) -> tuple[_KT, _VT]: ...
+    @overload  # type:ignore[override]
+    def update(self, __m: Collection[tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
+    @overload
+    def update(self, **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: SupportsItems[_KT, _VT], **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: SupportsKeysAndGetItem[_KT, _VT], **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: Iterable[tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: None = ..., **kwargs: _VT) -> None: ...
+    def __len__(self) -> int: ...
+    def keys(self) -> Iterator[_KT]: ...  # type:ignore[override]
+    __iter__ = keys
+    def values(self) -> Iterator[_VT]: ...  # type:ignore[override]
+    def items(self) -> Iterator[tuple[_KT, _VT]]: ...  # type:ignore[override]
+
+class GetDict(MultiDict[str, str]):
+    env: dict[str, Any]
+    @overload
+    def __init__(self, data: SupportsItems[str, str], env: dict[str, Any]) -> None: ...
+    @overload
+    def __init__(self, data: Iterable[tuple[str, str]], env: dict[str, Any]) -> None: ...
+    def on_change(self) -> None: ...
+    def copy(self) -> GetDict: ...
+
+class NestedMultiDict(MultiDict[_KT, _VT]):
+    dicts: tuple[MultiDict[_KT, _VT]]
+    def __init__(self, *dicts: MultiDict[_KT, _VT]) -> None: ...
+    def __getitem__(self, key): ...
+    def __setitem__(self, key: _KT, value: _VT) -> NoReturn: ...
+    def add(self, key: _KT, value: _VT) -> NoReturn: ...
+    def __delitem__(self, key: _KT) -> NoReturn: ...
+    def clear(self) -> NoReturn: ...
+    def setdefault(self, key: _KT, default: _VT | None = ...) -> NoReturn: ...
+    def pop(self, key: _KT, default: Any = ...) -> NoReturn: ...
+    def popitem(self) -> NoReturn: ...
+    def update(self, *args: Any, **kwargs: _VT) -> NoReturn: ...
+    def copy(self) -> NestedMultiDict[_KT, _VT]: ...
+    def __nonzero__(self) -> bool: ...
+
+class NoVars:
+    reason: str
+    def __init__(self, reason: str | None = ...) -> None: ...
+    def __getitem__(self, key) -> None: ...
+    def __setitem__(self, *args, **kw) -> None: ...
+    add = __setitem__
+    setdefault = __setitem__
+    update = __setitem__
+    def __delitem__(self, *args, **kw) -> None: ...
+    clear = __delitem__
+    pop = __delitem__
+    popitem = __delitem__
+    def get(self, key, default: Incomplete | None = ...): ...
+    def getall(self, key): ...
+    def getone(self, key): ...
+    def mixed(self): ...
+    dict_of_lists = mixed
+    def __contains__(self, key) -> bool: ...
+    has_key = __contains__
+    def copy(self): ...
+    def __len__(self) -> int: ...
+    def iterkeys(self): ...
+    keys = iterkeys
+    items = iterkeys
+    values = iterkeys
+    __iter__ = iterkeys
```

Comment:
I forgot to annotate these. This is mostly a note to myself. It would be nice if there was a way to ensure `MultiDict | NoVars` could narrow to just `MultiDict` with a simple truthyness check, since `NoVars` is never truthy. Unfortunately I don't think mypy does any type narrowing on static truthyness of types yet. It would be annoying to make everywhere accept either of these since they don't share a common superclass.

Reply:
Mypy can certainly use truthiness for type narrowing! https://mypy-play.net/?mypy=latest&python=3.11&gist=5322fe41b970986d96a392ab5e31f021

Reply:
Nice, in my short experiment it didn't seem to narrow down the type union by doing exactly what you did in your simple example. I will definitely follow up on this, when I play around with the stubs in some code bases, since `if request.POST:` is a common pattern when using WebOb and deciding whether or not to pass the `POST` data forward to something like `WTForms` which will behave differently depending on whether it was passed something empty or nothing at all.